Playlists: convert to use reconcilers instead

charandas opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Reconcilers have the advantage where the older object is not needed to be kept around for the events to be processed.

This changeset converts playlists (which is the template for future SDK generation for Grafana kinds) to use reconciler in anticipation that the codegen in SDK can follow this example.

The feature flag is definitely not in use anywhere, hence changing its name, instead of maintaining the older one.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

Grafana App Platform

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

charandas committed:
Playlists: convert to use reconcilers instead

IfSentient commented on apps/playlist/pkg/watchers/reconciler_playlist.go:
```diff
@@ -0,0 +1,62 @@
+			}
+			return operator.ReconcileResult{}, err
+		case operator.ReconcileActionUnknown:
+			return operator.ReconcileResult{}, errors.New("Invalid reconcile action unknown")
```
`simple.App` uses an `operator.InformerController`, [which will retry errors returned from the reconciler according to its RetryPolicy](https://github.com/grafana/grafana-app-sdk/blob/main/operator/informer_controller.go#L562-L570). By default this policy is an [ExponentialBackoffRetryPolicy which will do 5 total retries](https://github.com/grafana/grafana-app-sdk/blob/main/operator/informer_controller.go#L35). To avoid retrying errors like this which will not be resolved on a retry, we may want to either set up a custom RetryPolicy that doesn't retry these errors in [AppConfig.InformerConfig](https://github.com/grafana/grafana-app-sdk/blob/main/simple/app.go#L114) (perhaps use a type which implements `error` and some `Retryable() bool` function which can be checked in the RetryPolicy), or log the error and return `nil` so the ReconcileRequest won't be retried later.

charandas replied:
Thank you. Based on that, I will go with:

> log the error and return nil so the ReconcileRequest won't be retried later.

IfSentient commented on apps/playlist/pkg/watchers/reconciler_playlist.go:
```diff
@@ -0,0 +1,62 @@
+		case operator.ReconcileActionUnknown:
+			return operator.ReconcileResult{}, errors.New("Invalid reconcile action unknown")
+		}
+		return operator.ReconcileResult{}, errors.New("Invalid reconcile action")
```
Same comment here about errors being retried.

IfSentient reviewed (commented):
Overall looks good, just a note about how errors returned from the Reconcile call are handled by the InformerController.

radiohead commented on apps/playlist/pkg/watchers/reconciler_playlist.go:
```diff
@@ -0,0 +1,62 @@
+	playlist "github.com/grafana/grafana/apps/playlist/pkg/apis/playlist/v0alpha1"
+)
+
+func NewPlaylistReconciler() operator.Reconciler {
```
We do have the typed reconciler which uses generics – we should be able to use it here – https://github.com/grafana/grafana-app-sdk/blob/55093e19b720517beaee7891288a84f5f7805d17/operator/reconciler.go#L253-L259

charandas committed:
convert to typed reconciler - not working yet

charandas committed:
rebase from main

charandas committed:
fix the update on opinionated reconciler

charandas commented:
I have gone ahead and addressed all the review comments. Also, filed this in relation to this effort: https://github.com/grafana/grafana-app-sdk/issues/559

charandas committed:
remove redundant log

charandas committed:
remove redundant import
